cps/kni.c

```diff
@@ -613,6 +587,152 @@ mnl_socket_recvfrom_flags(const struct mnl_socket *nl, void *buf, size_t bufsiz,
+				gw_buf, sizeof(gw_buf)) == NULL)
+			return -1;
+	} else {
+		proto = ETHER_TYPE_IPv4;
+
+		prefix_fib_id = lpm_lookup_ipv4(ltbl->lpm, update->ip.v4.s_addr);
+		if (prefix_fib_id < 0) {
+			gw_fib_id = lpm_lookup_ipv4(
+				ltbl->lpm, update->gw.v4.s_addr);
+			if (gw_fib_id < 0)
+				return -1;
+			gw_fib = &ltbl->fib_tbl[gw_fib_id];
+		} else
+			prefix_fib = &ltbl->fib_tbl[prefix_fib_id];
+
+		if (inet_ntop(AF_INET, &update->ip.v4.s_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+
+		if (inet_ntop(AF_INET, &update->gw.v4.s_addr,
+				gw_buf, sizeof(gw_buf)) == NULL)
+			return -1;
+	}
+
+	if (prefix_fib && prefix_fib->action == GK_FWD_GRANTOR &&
+			prefix_fib->u.grantor.gt_addr.proto == proto)
+		gt_addr = &prefix_fib->u.grantor.gt_addr;
+	else if (prefix_fib && prefix_fib->action == GK_GET_GRANTOR &&
+			prefix_fib->u.grantor_ip.proto == proto)
+		gt_addr = &prefix_fib->u.grantor_ip;
+
+	if (gt_addr != NULL) {
+		if (proto == ETHER_TYPE_IPv4) {
+			if (inet_ntop(AF_INET, &gt_addr->ip.v4.s_addr,
+					gt_buf, sizeof(gt_buf)) == NULL)
+				return -1;
+		} else {
+			if (inet_ntop(AF_INET, &gt_addr->ip.v6.s6_addr,
+					gt_buf, sizeof(gt_buf)) == NULL)
+				return -1;
+		}
+
+		return add_fib_entry(ipp_buf, gt_buf, gw_buf,
+			GK_FWD_GRANTOR, gk_conf);
+	}
+
+	if ((gw_fib && gw_fib->action == GK_FWD_NEIGHBOR_FRONT_NET) ||
+			(prefix_fib && prefix_fib->action ==
+			GK_FWD_GATEWAY_FRONT_NET)) {
+		return add_fib_entry(ipp_buf, NULL, gw_buf,
+			GK_FWD_GATEWAY_FRONT_NET, gk_conf);
+	}
+
+	if ((gw_fib && gw_fib->action == GK_FWD_NEIGHBOR_BACK_NET) ||
+			(prefix_fib && prefix_fib->action ==
+			GK_FWD_GATEWAY_BACK_NET)) {
+		return add_fib_entry(ipp_buf, NULL, gw_buf,
+			GK_FWD_GATEWAY_BACK_NET, gk_conf);
+	}
+
+	return -1;
+}
+
+static int
+del_route(struct route_update *update, struct gk_config *gk_conf)
+{
+	int ret;
+	char ip_buf[INET6_ADDRSTRLEN];
+	char ipp_buf[INET6_ADDRSTRLEN + 4];
+
+	if (update->family == AF_INET6) {
+		if (inet_ntop(AF_INET6, &update->ip.v6.s6_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+	} else {
+		if (inet_ntop(AF_INET, &update->ip.v4.s_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+	}
+
+	return del_fib_entry(ipp_buf, gk_conf);
+}
+
+static inline void
+update_gk_lpm(struct route_update *updates,
+	unsigned int num_updates, struct gk_config *gk_conf)
+{
+	unsigned int i;
+	for (i = 0; i < num_updates; i++) {
+		if (updates[i].type == RTM_NEWROUTE)
+			new_route(&updates[i], gk_conf);
+		else
+			del_route(&updates[i], gk_conf);
+	}
+}
```

Comment:
Is there a reason to add this code as its own function? Since it is just a for loop and there is only one use of it in the code.